File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        let prev = null;
        let cur = head
        let next = head?.next;

        while (cur) {
            cur.next = prev;
            prev = cur;
            cur = next;
            next = next?.next
        }

        return prev;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) return null;
        if (head.next == null && n == 1) return null;

        var dummy = new ListNode(-1, head);
        removeNthFromEndRec(dummy, n);
        
        return dummy.next;
    }

    private int removeNthFromEndRec(ListNode head, int n) {
        if (head == null) return n;

        var counter = removeNthFromEndRec(head.next, n);
        if (counter == 0) {
            ListNode node = null;
            if (head.next != null) {
                node = head.next.next;
            }
            head.next = node;
        }

        return counter - 1;
    }
}
